cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +419,259 @@ class self_comparator {
+          r_end_off   = r_list_col.offset_at(r_end_off);
+          if (l_end_off - l_start_off != r_end_off - r_start_off) { return false; }
+        }
+      }
+
+      for (int i = l_start_off, j = r_start_off; i < l_end_off; ++i, ++j) {
+        bool equal = type_dispatcher<dispatch_void_if_nested>(
+          lcol.type(), element_comparator{nulls, lcol, rcol, nulls_are_equal}, i, j);
+        if (not equal) { return false; }
+      }
+      return true;
+    }
+
+   private:
+    column_device_view const lhs;
+    column_device_view const rhs;
+    Nullate const nulls;
+    null_equality const nulls_are_equal;
+  };
+
+  table_device_view const lhs;
+  table_device_view const rhs;
+  Nullate const nulls;
+  null_equality const nulls_are_equal;
+};
+
+struct preprocessed_table {
+  /**
+   * @brief Preprocess table for use with row equality comparison or row hashing
+   *
+   * Sets up the table for use with row equality comparison or row hashing. The resulting
+   * preprocessed table can be passed to the constructor of `equality::self_comparator` to
+   * avoid preprocessing again.
+   *
+   * @param table The table to preprocess
+   * @param stream The cuda stream to use while preprocessing.
+   */
+  static std::shared_ptr<preprocessed_table> create(table_view const& table,
+                                                    rmm::cuda_stream_view stream);
+
+ private:
+  friend class self_comparator;
+
+  using table_device_view_owner =
+    std::invoke_result_t<decltype(table_device_view::create), table_view, rmm::cuda_stream_view>;
+
+  preprocessed_table(table_device_view_owner&& table,
+                     std::vector<rmm::device_buffer>&& null_buffers)
+    : _t(std::move(table)), _null_buffers(std::move(null_buffers))
+  {
+  }
+
+  /**
+   * @brief Implicit conversion operator to a `table_device_view` of the preprocessed table.
+   *
+   * @return table_device_view
+   */
+  operator table_device_view() { return *_t; }
+
+  table_device_view_owner _t;
+  std::vector<rmm::device_buffer> _null_buffers;
+};
+
+class self_comparator {
+ public:
+  /**
+   * @brief Construct an owning object for performing equality comparisons between two rows of the
+   * same table.
+   *
+   * @param t The table to compare
+   * @param stream The stream to construct this object on. Not the stream that will be used for
+   * comparisons using this object.
+   */
+  self_comparator(table_view const& t, rmm::cuda_stream_view stream)
+    : d_t(preprocessed_table::create(t, stream))
+  {
+  }
+
+  /**
+   * @brief Construct an owning object for performing equality comparisons between two rows of the
+   * same table.
+   *
+   * This constructor allows independently constructing a `preprocessed_table` and sharing it among
+   * multiple comparators.
+   *
+   * @param t A table preprocessed for equality comparison
+   */
+  self_comparator(std::shared_ptr<preprocessed_table> t) : d_t{std::move(t)} {}
```

Comment:
Apologies if we already discussed this on the struct PR: why do we want to move `t`? The whole point of `preprocessed_table::create` is so that a user can make multiple comparators referring to the same table if they want, right? Doesn't moving destroy the possibility of sharing?

Reply:
It eliminates a redundant copy.

A copy of the `shared_ptr` is made when passing by value to `create`. Then `d_t{std::move(t)}` will move that copy into the `d_t` member. Without the `move`, `d_t{t}` would make an _additional_ copy of the `shared_ptr`. Then the copy that was created for the function parameter passed by value would be destroyed.

Reply:
Ah of course the parameter is passed by value so we're making a copy. Got it, thanks.